### PR TITLE
ENH: dynamic class magic for twincat state config PVs

### PR DESCRIPTION
**Pull request by ZLLentz**
## Description

Replace the `TwinCATStateConfigAll` class with a `TwinCATStateConfigDynamic` class that produces a different dynamically generated subclass based on the parent device's class attributes.

Increases the max number of allowable state config elements from 5 to 9.

Configure all the twincat state devices for the number of states that they use on the beamline.

Add a special class for TM2K2, which has one more state than the standard ATMs.

## Motivation and Context

This addresses two issues:
- Some classes aren't currently showing all the relevant configuration PVs. https://jira.slac.stanford.edu/browse/LCLSPC-139
- Some classes show too many configuration PVs, making it confusing to find the ones that are relevant.

## How Has This Been Tested?

Added relevant unit tests. The devices I changed still pass their tests. The devices I've tried in typhos look great.

Here's IM3L0, a PPM. This is what you get when you switch to the tree template for the states device (you do not see the nested config parameters at all with the normal template)

image

## Where Has This Been Documented?

Added pre-release notes and docstrings. We'll wait for the tag to re-run the api script to avoid annoying merge conflicts.

## Pre-merge checklist
- [x] Code works interactively
- [x] Code contains descriptive docstrings, including context and API
- [x] New/changed functions and methods are covered in the test suite where possible
- [x] Test suite passes locally
- [x] Test suite passes on travis
- [x] Ran docs/pre-release-notes.sh and created a pre-release documentation page
- [x] Pre-release docs include context, functional descriptions, and contributors as appropriate

### CODE REVIEW

**Comment by ZLLentz** [comment]
I also think I want to change the implementation a bit here: I'm starting to dislike magic class variables, and will check if there's a way to implement this entirely using `__init__` parameters passed through `Component`/`UpdateComponent`. I think that will make it much more clear what's going on when reading the code.

**Comment by klauer** [comment]
Something like redefining `config` on the subclass, with the appropriate number?
```
    config = Cpt(TwinCATStateConfigDynamic(states=8), '', kind='omitted',
                 doc='Configuration of state positions, deltas, etc.')
```

**Comment by ZLLentz** [comment]
> AT1K4 and AT2K2 also need fixing in this PR (10 states, IIRC?)

The motion library currently has max states set to 9: https://github.com/pcdshub/lcls-twincat-motion/blob/b06601abf7178475651e279f94becbe4ec9926bf/lcls-twincat-motion/Library/GVLs/MOTION_GVL.TcGVL#L11

Is this causing any sort of issue already?

> Something like redefining config on the subclass, with the appropriate number?

Yes, precisely this- but hopefully the `UpdateComponent` makes this simple.

**Comment by klauer** [comment]
AT1K4 - Off by 1, I was including 'unknown' and you aren't, I believe.

**Comment by ZLLentz** [comment]
There were some subtle bugs related to copy vs deepcopy, recursion depths, and managing fake dynamic classes. I think I've taken care of them all now (or, at least the tests pass locally). There was much rejoicing. :tada:

Tomorrow I'm going to clean up annotations, docstrings, and maybe some oddly-chosen function names.

**Comment by ZLLentz** [comment]
My refactor to lose the magic class variables is definitely more verbose than the previous version, but I think it's easier to understand and is much more likely to throw an exception when you make a typo.

**Review by klauer — APPROVED** [review]
Wow - nicely done. Not that it matters for the review, really, but I would not have come up with as clean of a solution for such a messy problem.

Verbose docstrings, explanations, and comments made it easy to read through.

Not much more to say than that. 👍

File: docs/source/upcoming_release_notes/875-twincat-states.rst [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+  of state configuration PVs. These are the structures that allow you to
+  check and change state setpoints, deltas, velocities, etc. This is
+  implemented through the new TwinCATStateDynamic class.
+- Increase the maximum number of connected state configuration records to
+  match the current motion library limit (9)
+
+Device Updates
+--------------
+- Using the new TwinCATStateDynamic mechanisms and the UpdateComponent,
+  Update the following classes to contain exactly the correct number of
```
[the remark this answers is not in the document]
Comment (ZLLentz) [reply docs/source/upcoming_release_notes/875-twincat-states.rst]: I'll add this line in some form, I think this diff as-is breaks the sentence flow after my other recent edits

File: pcdsdevices/device.py [hunk 2]
```diff
@@ -331,14 +342,18 @@ def __set_name__(self, owner, attr_name):
         # Defer to the normal component setup for the rest
         super().__set_name__(owner, attr_name)
 
-    def __getattr__(self, name):
+    def __getattr__(self, name: str):
         # If we are missing something, check the copied/edited hostage cpt
         return getattr(self.copy_cpt, name)
```
[the remark this answers is not in the document]
Comment (ZLLentz) [reply pcdsdevices/device.py:347]: I think this is correct if we want to enforce the immutability of the UpdateComponent, I'll try it and make sure it doesn't break the test suite and we weren't relying on this accidental mutability somewhere.

File: pcdsdevices/state.py [hunk 3]
```diff
@@ -497,19 +506,119 @@ class TwinCATStateConfigOne(Device):
...
+                    Cpt(
+                        cls._config_cls,
+                        f':{num:02}',
+                        kind='config',
+                    )
+                    for num in range(1, state_count + 1)
+                }
+            )
+            cls._state_config_registry[state_count] = new_cls
+        return super().__new__(new_cls)
```
[the remark this answers is not in the document]
Comment (ZLLentz) [reply pcdsdevices/state.py:549]: I think this is a good pattern to be aware of and it seems to have good results.

File: pcdsdevices/state.py [hunk 4]
```diff
@@ -572,12 +673,32 @@ class TwinCATStatePositioner(StatePositioner):
...
+    )
 
     set_metadata(error_id, dict(variety='scalar', display_format='hex'))
     set_metadata(reset_cmd, dict(variety='command', value=1))
 
+    def __init_subclass__(cls, **kwargs):
+        # We need to adjust the state enum_attrs appropriately if
+        # state_count was updated.
+        state_count = cls.config.kwargs['state_count']
+        parent_count = cls.mro()[1].config.kwargs['state_count']
```
[the remark this answers is not in the document]
Comment (ZLLentz) [reply pcdsdevices/state.py:691]: I'll check, you're probably right.

[the remark this answers is not in the document]
Comment (ZLLentz) [reply pcdsdevices/state.py:691]: It looks like `UpdateComponent` retains no references to the original component- should it?

In general, I don't think this code here should assume that the `UpdateComponent` utility was used. I think it should work the same regardless on how the component was specified, so long as a `state_count` was passed (maybe I should have a guard for a mistake where something without a `state_count` sneaks into the config component?)

Not 100% sure on all of this

[the remark this answers is not in the document]
Comment (klauer) [reply pcdsdevices/state.py:695]: I think it's fine as-is. 👍

[the remark this answers is not in the document]
Comment (ZLLentz) [reply pcdsdevices/state.py:695]: As an aside, I wish `super()` would work as a substitution for `cls.mro()[1]` in class methods.

**Pull request merged**